common/persistence/nosql/shardedNosqlStore.go

```diff
@@ -0,0 +1,147 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package nosql
+
+import (
+	"fmt"
+	"sync"
+
+	p "github.com/uber/cadence/common/persistence"
+
+	"github.com/uber/cadence/common/config"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+)
+
+type (
+	// shardedNosqlStore is a store that may have one or more shards
+	shardedNosqlStore struct {
+		sync.RWMutex
+
+		config config.ShardedNoSQL
+		dc     *p.DynamicConfiguration
+		logger log.Logger
+
+		connectedShards map[string]nosqlStore
+		defaultShard    nosqlStore
+		shardingPolicy  shardingPolicy
+	}
+)
+
+func newShardedNosqlStore(cfg config.ShardedNoSQL, logger log.Logger, dc *p.DynamicConfiguration) (*shardedNosqlStore, error) {
+	sn := shardedNosqlStore{
+		config: cfg,
+		dc:     dc,
+		logger: logger,
+	}
+
+	// Connect to the default shard
+	defaultShardName := cfg.DefaultShard
+	store, err := sn.connectToShard(cfg.Connections[defaultShardName].NoSQLPlugin, defaultShardName)
+	if err != nil {
+		return nil, err
+	}
+	sn.defaultShard = *store
+	sn.connectedShards = map[string]nosqlStore{
+		defaultShardName: sn.defaultShard,
+	}
+
+	// Parse & validate the sharding policy
+	sn.shardingPolicy, err = newShardingPolicy(logger, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sn, nil
+}
+
+func (sn *shardedNosqlStore) GetStoreShardByHistoryShard(shardID int) (*nosqlStore, error) {
+	shardName := sn.shardingPolicy.getHistoryShardName(shardID)
+	return sn.getShard(shardName)
+}
+
+func (sn *shardedNosqlStore) GetStoreShardByTaskList(domainID string, taskListName string, taskType int) (*nosqlStore, error) {
+	shardName := sn.shardingPolicy.getTaskListShardName(domainID, taskListName, taskType)
+	return sn.getShard(shardName)
+}
+
+func (sn *shardedNosqlStore) GetDefaultShard() nosqlStore {
+	return sn.defaultShard
+}
+
+func (sn *shardedNosqlStore) Close() {
+	sn.defaultShard.Close()
+
+	for name, shard := range sn.connectedShards {
+		sn.logger.Warn("Closing store shard", tag.StoreShard(name))
+		shard.Close()
+	}
+}
+
+func (sn *shardedNosqlStore) GetName() string {
+	return "shardedNosql"
+}
+
+func (sn *shardedNosqlStore) getShard(shardName string) (*nosqlStore, error) {
+	sn.RLock()
+	shard, found := sn.connectedShards[shardName]
+	sn.RUnlock()
+
+	if found {
+		return &shard, nil
+	}
+
+	cfg, ok := sn.config.Connections[shardName]
+	if !ok {
+		return nil, &ShardingError{
+			Message: fmt.Sprintf("Unknown db shard name: %v", shardName),
+		}
+	}
+
+	sn.Lock()
+	if shard, ok := sn.connectedShards[shardName]; ok { // read again to double-check
+		sn.Unlock()
+		return &shard, nil
+	}
+
+	s, err := sn.connectToShard(cfg.NoSQLPlugin, shardName)
```

Comment:
if we don't have any already (and I may have missed them) it'd be great to get some metrics around the number of connections occurring